File: zig-generator/src/emitters/ldr_reg.zig
const std = @import("std");
const Emitter = @import("../emitter.zig").Emitter;
const GeneratorState = @import("../state.zig").GeneratorState;

pub const LdrRegEmitter = struct {
    allocator: std.mem.Allocator,

    pub fn emitter(self: *LdrRegEmitter) Emitter {
        return .{ .ptr = self, .vtable = &.{ .try_emit = tryEmit } };
    }

    fn tryEmit(ptr: *anyopaque, state: *GeneratorState) ?[]u8 {
        const self: *LdrRegEmitter = @ptrCast(@alignCast(ptr));
        const scratch_id = state.getScratchRegionId() orelse return null;
        const scratch = state.getRegion(scratch_id);
        if (scratch.watermark < 8) return null;
        const dst    = state.pickWritableGpr() orelse return null;
        const offset = state.getRng().uintLessThan(u64, scratch.watermark / 8) * 8;
        state.markScalar(dst);
        return std.fmt.allocPrint(self.allocator, "  LDR x{d}, [x22, #{d}]", .{ dst, offset }) catch null;
    }
};

File: rust-generator/src/emitters/ldr_reg.rs
use crate::emitter::Emitter;
use crate::state::GeneratorState;
use rand::RngExt;

pub struct LdrRegEmitter;

impl Emitter for LdrRegEmitter{
    fn try_emit(&mut self, state: &mut GeneratorState) -> Option<String> {
        let scratch_id = state.get_scratch_region_id()?;
        let watermark = state.get_region(scratch_id).watermark;
        let dst = state.pick_writable_gpr()?;
        let offset = state.rng().random_range(0..watermark / 8) * 8;
        state.mark_scalar(dst);


        Some(format!("  LDR x{}, [x22, #{}]", dst, offset))
    }
}
